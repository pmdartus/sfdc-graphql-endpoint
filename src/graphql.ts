import {
    GraphQLBoolean,
    GraphQLEnumType,
    GraphQLFieldConfig,
    GraphQLFieldConfigArgumentMap,
    GraphQLFieldResolver,
    GraphQLFloat,
    GraphQLID,
    GraphQLInputFieldConfig,
    GraphQLInputObjectType,
    GraphQLInputType,
    GraphQLInt,
    GraphQLInterfaceType,
    GraphQLList,
    GraphQLNonNull,
    GraphQLObjectType,
    GraphQLOutputType,
    GraphQLScalarType,
    GraphQLSchema,
    GraphQLString,
} from 'graphql';

import {
    ChildRelationship,
    Entity,
    Field,
    FieldType,
    isPolymorphicReference,
    isReferenceField,
    isScalarField,
    ReferenceField,
    ScalarField,
    SfdcSchema,
} from './sfdc/schema.js';

interface SchemaConfig<Context> {
    sfdcSchema: SfdcSchema;
    resolvers?: {
        query?: (entity: Entity, schema: SfdcSchema) => GraphQLFieldResolver<unknown, Context>;
        queryMany?: (entity: Entity, schema: SfdcSchema) => GraphQLFieldResolver<unknown, Context>;
    };
}

interface State {
    config: SchemaConfig<any>;

    /** Map associating SObject names to GraphQL object types. */
    entityTypes: Map<string, GraphQLObjectType>;
    /** Map associating SObject names to GraphQL where predicate object types. */
    whereTypes: Map<string, GraphQLInputObjectType>;
    /** Map associating SObject names to GraphQL order by object types. */
    orderTypes: Map<string, GraphQLList<GraphQLInputObjectType>>;
}

const SCALAR_TYPES_MAPPING: { [type in ScalarField['type']]: GraphQLScalarType } = {
    // Builtin GraphQL scalars
    [FieldType.ID]: GraphQLID,
    [FieldType.STRING]: GraphQLString,
    [FieldType.BOOLEAN]: GraphQLBoolean,
    [FieldType.INT]: GraphQLInt,
    [FieldType.FLOAT]: GraphQLFloat,

    // Custom SFDC types
    [FieldType.ENCRYPTED_STRING]: new GraphQLScalarType({ name: 'EncryptedString' }),
    [FieldType.DATE]: new GraphQLScalarType({ name: 'Date' }),
    [FieldType.DATETIME]: new GraphQLScalarType({ name: 'DateTime' }),
    [FieldType.TIME]: new GraphQLScalarType({ name: 'Time' }),
    [FieldType.BASE64]: new GraphQLScalarType({ name: 'Base64' }),
    [FieldType.CURRENCY]: new GraphQLScalarType({ name: 'Currency' }),
    [FieldType.TEXTAREA]: new GraphQLScalarType({ name: 'TextArea' }),
    [FieldType.PERCENT]: new GraphQLScalarType({ name: 'Percent' }),
    [FieldType.PHONE]: new GraphQLScalarType({ name: 'Phone' }),
    [FieldType.URL]: new GraphQLScalarType({ name: 'URL' }),
    [FieldType.EMAIL]: new GraphQLScalarType({ name: 'Email' }),
    [FieldType.COMBOBOX]: new GraphQLScalarType({ name: 'Combobox' }),
    [FieldType.ANY_TYPE]: new GraphQLScalarType({ name: 'AnyType' }),
    [FieldType.PICKLIST]: new GraphQLScalarType({ name: 'Picklist' }),
    [FieldType.MULTI_PICKLIST]: new GraphQLScalarType({ name: 'MultiPicklist' }),
    [FieldType.ADDRESS]: new GraphQLScalarType({ name: 'Address' }),
    [FieldType.LOCATION]: new GraphQLScalarType({ name: 'Location' }),
};

const ENTITY_INTERFACE = new GraphQLInterfaceType({
    name: 'Entity',
    fields: {
        Id: {
            type: new GraphQLNonNull(GraphQLID),
        },
        Name: {
            type: GraphQLString,
        },
        CreatedDate: {
            type: new GraphQLNonNull(SCALAR_TYPES_MAPPING[FieldType.DATETIME]),
        },
        LastModifiedDate: {
            type: new GraphQLNonNull(SCALAR_TYPES_MAPPING[FieldType.DATETIME]),
        },
        SystemModstamp: {
            type: new GraphQLNonNull(SCALAR_TYPES_MAPPING[FieldType.DATETIME]),
        },
        LastViewedDate: {
            type: SCALAR_TYPES_MAPPING[FieldType.DATETIME],
        },
    },
});

const OPERATOR_INPUT_TYPES_MAPPING = Object.fromEntries(
    Object.entries(SCALAR_TYPES_MAPPING).map(([key, type]) => [key, createInputOperator({ type })]),
) as { [type in ScalarField['type']]: GraphQLInputObjectType };

export type GraphQLSortOrderValue =
    | 'ASC'
    | 'DESC'
    | 'ASC_NULLS_FIRST'
    | 'ASC_NULLS_LAST'
    | 'DESC_NULLS_FIRST'
    | 'DESC_NULLS_LAST';

const ORDER_BY_ENUM_TYPE_VALUES: { [name in GraphQLSortOrderValue]: object } = {
    ASC: {},
    DESC: {},
    ASC_NULLS_FIRST: {},
    ASC_NULLS_LAST: {},
    DESC_NULLS_FIRST: {},
    DESC_NULLS_LAST: {},
};

const ORDER_BY_ENUM_TYPE = new GraphQLEnumType({
    name: 'OrderByEnum',
    values: ORDER_BY_ENUM_TYPE_VALUES,
});

export function entitiesToSchema<Context>(config: SchemaConfig<Context>): GraphQLSchema {
    const state: State = {
        config,
        entityTypes: new Map(),
        orderTypes: new Map(),
        whereTypes: new Map(),
    };

    for (const [name, entity] of Object.entries(config.sfdcSchema.entities)) {
        const type = createGraphQLEntityType(state, entity);
        state.entityTypes.set(name, type);
    }

    const query = createQuery(state);

    return new GraphQLSchema({
        query,
        types: Array.from(state.entityTypes.values()),
    });
}

function createGraphQLEntityType(state: State, entity: Entity): GraphQLObjectType {
    const { name, fields, childRelationships } = entity;

    return new GraphQLObjectType({
        name,
        interfaces: [ENTITY_INTERFACE],
        fields: () => {
            const graphQLFields = fields.map((field) => [
                field.name,
                createGraphQLEntityField(state, field),
            ]);
            const graphQLRelationships = childRelationships
                .map((relationship) => [
                    relationship.name,
                    createGraphQLEntityRelationships(state, relationship),
                ])
                .filter(([, value]) => value !== undefined);

            return Object.fromEntries([...graphQLFields, ...graphQLRelationships]);
        },
    });
}

function createGraphQLEntityField(
    state: State,
    field: Field,
): GraphQLFieldConfig<unknown, unknown> {
    let type: GraphQLOutputType;
    let resolve: GraphQLFieldResolver<any, unknown> = (source) => {
        return source[field.name];
    };

    if (isScalarField(field)) {
        type = SCALAR_TYPES_MAPPING[field.type];
    } else if (isReferenceField(field)) {
        const { referencedEntity } = field;

        if (!referencedEntity) {
            type = GraphQLID;
        } else {
            type = state.entityTypes.get(referencedEntity.name)!;
            resolve = (source) => {
                return source[field.relationshipName];
            };
        }
    } else {
        // TODO: Add support for polymorphic field lookups.
        type = GraphQLID;
    }

    if (!field.config.nillable) {
        type = new GraphQLNonNull(type);
    }

    return {
        type,
        resolve,
    };
}

function createGraphQLEntityRelationships(
    state: State,
    relationship: ChildRelationship,
): GraphQLFieldConfig<unknown, unknown> | undefined {
    const { entity } = relationship;

    // Ignore all the children relationships that aren't part of the SFDC graph. While it's possible
    // to produce all the children relationships, it would bloat the schema.
    if (!entity) {
        return;
    }

    const type = state.entityTypes.get(entity.name)!;
    const args = createEntityListArgs(state, entity);

    return {
        args,
        type: new GraphQLList(type),
    };
}

function createQuery(state: State): GraphQLObjectType {
    const { sfdcSchema } = state.config;
    const entities = Object.values(sfdcSchema.entities);

    return new GraphQLObjectType({
        name: 'Query',
        fields: () => {
            return Object.assign(
                {},
                ...entities.map((entity) => createEntityQueries(state, entity)),
            );
        },
    });
}

function createEntityQueries(
    state: State,
    entity: Entity,
): Record<string, GraphQLFieldConfig<unknown, unknown>> {
    const { name } = entity;
    const { resolvers, sfdcSchema } = state.config;

    const type = state.entityTypes.get(name)!;

    const queryManyArgs = createEntityListArgs(state, entity);
    const querySingleArgs = {
        id: {
            type: GraphQLID,
        },
    };

    return {
        [name]: {
            args: queryManyArgs,
            type: new GraphQLList(type),
            resolve: resolvers?.queryMany?.(entity, sfdcSchema),
        },
        [`${name}_by_id`]: {
            args: querySingleArgs,
            type: type,
            resolve: resolvers?.query?.(entity, sfdcSchema),
        },
    };
}

function createEntityListArgs(state: State, entity: Entity): GraphQLFieldConfigArgumentMap {
    const whereInputType = getWhereInputType(state, entity);
    const orderByInputType = getOderByInputType(state, entity);

    return {
        limit: {
            type: new GraphQLNonNull(GraphQLInt),
        },
        offset: {
            type: GraphQLInt,
        },
        where: {
            type: whereInputType,
        },
        order_by: {
            type: orderByInputType,
        },
    };
}

function getOderByInputType(state: State, entity: Entity): GraphQLInputType {
    const { name, fields } = entity;
    let orderByInputType = state.orderTypes.get(name);

    if (orderByInputType === undefined) {
        orderByInputType = new GraphQLList(
            new GraphQLInputObjectType({
                name: `${name}OrderBy`,
                fields: () =>
                    Object.fromEntries(
                        fields
                            .filter((field) => field.config.sortable)
                            .map((field) => {
                                let type: GraphQLInputType = ORDER_BY_ENUM_TYPE;
                                if (isReferenceField(field) && field.referencedEntity) {
                                    type = getOderByInputType(state, field.referencedEntity);
                                }

                                return [
                                    field.name,
                                    {
                                        type,
                                    },
                                ];
                            }),
                    ),
            }),
        );

        state.orderTypes.set(name, orderByInputType);
    }

    return orderByInputType;
}

function getWhereInputType(state: State, entity: Entity): GraphQLInputType {
    const { name, fields } = entity;
    let whereInputType = state.whereTypes.get(name);

    if (whereInputType === undefined) {
        // The type has to be explicitly set here to please TypeScript, otherwise it bails out
        // because of the recursive reference.
        const _whereInputType: GraphQLInputObjectType = new GraphQLInputObjectType({
            name: `${name}Where`,
            fields: () => {
                return {
                    _and: {
                        type: new GraphQLList(_whereInputType),
                    },
                    _or: {
                        type: new GraphQLList(_whereInputType),
                    },
                    ...Object.fromEntries(
                        fields
                            .filter(
                                (field): field is ScalarField | ReferenceField =>
                                    field.config.filterable && !isPolymorphicReference(field),
                            )
                            .map((field) => {
                                let type: GraphQLInputType;

                                if (isScalarField(field)) {
                                    type = OPERATOR_INPUT_TYPES_MAPPING[field.type];
                                } else if (isReferenceField(field)) {
                                    if (field.referencedEntity) {
                                        type = getWhereInputType(state, field.referencedEntity);
                                    } else {
                                        type = ORDER_BY_ENUM_TYPE;
                                    }
                                }

                                return [
                                    field.name,
                                    {
                                        type: type!,
                                    },
                                ];
                            }),
                    ),
                };
            },
        });

        whereInputType = _whereInputType;
        state.whereTypes.set(name, _whereInputType);
    }

    return whereInputType;
}

function createInputOperator({ type }: { type: GraphQLScalarType }): GraphQLInputObjectType {
    const { name } = type;
    const fields: Record<string, GraphQLInputFieldConfig> = {
        _eq: { type },
        _neq: { type },
        _gt: { type },
        _lt: { type },
        _gte: { type },
        _lte: { type },
        _in: { type: new GraphQLList(type) },
        _nin: { type: new GraphQLList(type) },
    };

    if (name === 'MultiPicklist') {
        fields._includes = { type };
        fields._excludes = { type };
    } else if (name === 'String') {
        fields._like = { type };
    }

    return new GraphQLInputObjectType({
        name: `${name}Operator`,
        fields,
    });
}
