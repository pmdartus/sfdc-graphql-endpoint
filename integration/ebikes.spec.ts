import { FastifyInstance } from 'fastify';
import { buildClientSchema, getIntrospectionQuery, printSchema } from 'graphql';

import { gql, getApp, executeQuery } from './utils.js';

let app: FastifyInstance;
beforeAll(() => {
    app = getApp({
        entities: ['Product__c', 'Product_Family__c', 'Order_Item__c', 'Order__c'],
    });
});

describe('eBikes entity', () => {
    describe('introspection', () => {
        test('introspect schema', async () => {
            const { data } = await executeQuery({
                app,
                query: getIntrospectionQuery({
                    schemaDescription: true,
                }),
            });

            const schema = buildClientSchema(data);
            expect(printSchema(schema)).toMatchSnapshot();
        });
    });

    describe('get product list', () => {
        test('retrieve most expensive products', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(order_by: { Price__c: DESC }, limit: 4) {
                            Name
                            Price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products order by gender and price', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(order_by: { Gender__c: ASC, Price__c: ASC }, limit: 10) {
                            Name
                            Price__c
                            Gender__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve more products', async () => {
            const query = gql`
                query getProducts($offset: Int = 0) {
                    Product__c(order_by: { Price__c: DESC }, limit: 2, offset: $offset) {
                        Name
                        Price__c
                    }
                }
            `;

            const { data: initialProducts } = await executeQuery({
                app,
                query,
                variables: {
                    offset: 0,
                },
            });
            expect(initialProducts).toMatchSnapshot();

            const { data: moreProducts } = await executeQuery({
                app,
                query,
                variables: {
                    offset: 2,
                },
            });
            expect(moreProducts).toMatchSnapshot();
        });

        test('retrieve the products with suspensions', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Suspension__c: { _eq: true } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                            Suspension__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products in price range', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Price__c: { _gt: 1300, _lt: 1500 } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products by exact Name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Name: { _eq: "Neomov - Basic" } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products by matching Name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Name: { _like: "Neomov%" } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products male or female products only', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Gender__c: { _in: ["Male", "Female"] } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                            Gender__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve the products with the "Rolling Mountain" family Name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { Product_Family__c: { Name: { _eq: "Rolling Mountain" } } }
                            order_by: { Name: ASC }
                            limit: 10
                        ) {
                            Name
                            Price__c
                            Product_Family__c {
                                Name
                            }
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });
    });

    describe('get single product', () => {
        test('retrieve a specific product by id', async () => {
            const {
                data: { Product__c: productList },
            } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(limit: 1) {
                            Id
                            Name
                        }
                    }
                `,
            });
            const {
                data: { Product__c_by_id: product },
            } = await executeQuery({
                app,
                query: gql`
                    query getProductById($id: ID!) {
                        Product__c_by_id(id: $id) {
                            Id
                            Name
                        }
                    }
                `,
                variables: {
                    id: productList[0].Id,
                },
            });

            expect(product).toEqual(productList[0]);
        });
    });

    describe('get orders', () => {
        test('retrieve products associated with order items', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Order__c(limit: 10) {
                            Order_Items__r(limit: 10, order_by: { Name: DESC }) {
                                Product__c {
                                    Name
                                    Price__c
                                }
                            }
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });
    });

    describe('SOQL query explanation', () => {
        test('Setting the "X-Explain-SOQL" should return planning explanation', async () => {
            const res = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(order_by: { Price__c: DESC }, limit: 4) {
                            Name
                            Price__c
                        }
                    }
                `,
                headers: {
                    'X-Explain-SOQL': 'true',
                }
            });

            expect(Array.isArray(res.data.Product__c)).toBe(true);
            expect(res.extensions).toMatchObject({
                explain: {
                    plans: expect.any(Array),
                    sourceQuery: expect.any(String),
                }
            })
        });
    });
});
