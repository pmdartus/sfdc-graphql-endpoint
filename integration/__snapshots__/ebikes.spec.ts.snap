// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eBikes entity get orders retrieve products associated with order items 1`] = `
Object {
  "Order__c": Array [
    Object {
      "Order_Items__r": Array [
        Object {
          "Product__c": Object {
            "Name": "Rolling Mountain - Male",
            "Price__c": 2300,
          },
        },
        Object {
          "Product__c": Object {
            "Name": "Neomov - Basic",
            "Price__c": 1450,
          },
        },
        Object {
          "Product__c": Object {
            "Name": "Wisper foldable - Upgraded",
            "Price__c": 1230,
          },
        },
      ],
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve more products 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
    },
    Object {
      "Name": "Rolling Mountain - Male",
      "Price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve more products 2`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Neomov - Suspension",
      "Price__c": 1850,
    },
    Object {
      "Name": "Neomov - Female frame",
      "Price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve most expensive products 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
    },
    Object {
      "Name": "Rolling Mountain - Male",
      "Price__c": 2300,
    },
    Object {
      "Name": "Neomov - Suspension",
      "Price__c": 1850,
    },
    Object {
      "Name": "Neomov - Male frame",
      "Price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products by exact Name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Neomov - Basic",
      "Price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products by matching Name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Neomov - Basic",
      "Price__c": 1450,
    },
    Object {
      "Name": "Neomov - Female frame",
      "Price__c": 1450,
    },
    Object {
      "Name": "Neomov - Male frame",
      "Price__c": 1450,
    },
    Object {
      "Name": "Neomov - Suspension",
      "Price__c": 1850,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products in price range 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Neomov - Basic",
      "Price__c": 1450,
    },
    Object {
      "Name": "Neomov - Female frame",
      "Price__c": 1450,
    },
    Object {
      "Name": "Neomov - Male frame",
      "Price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products male or female products only 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Gender__c": "Female",
      "Name": "Neomov - Female frame",
      "Price__c": 1450,
    },
    Object {
      "Gender__c": "Male",
      "Name": "Neomov - Male frame",
      "Price__c": 1450,
    },
    Object {
      "Gender__c": "Female",
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
    },
    Object {
      "Gender__c": "Male",
      "Name": "Rolling Mountain - Male",
      "Price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products order by gender and price 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Gender__c": "Unisex",
      "Name": "Wisper foldable - Upgraded",
      "Price__c": 1230,
    },
    Object {
      "Gender__c": "Unisex",
      "Name": "Wisper foldable",
      "Price__c": 1230,
    },
    Object {
      "Gender__c": "Unisex",
      "Name": "Neomov - Basic",
      "Price__c": 1450,
    },
    Object {
      "Gender__c": "Unisex",
      "Name": "Neomov - Suspension",
      "Price__c": 1850,
    },
    Object {
      "Gender__c": "Male",
      "Name": "Neomov - Male frame",
      "Price__c": 1450,
    },
    Object {
      "Gender__c": "Male",
      "Name": "Rolling Mountain - Male",
      "Price__c": 2300,
    },
    Object {
      "Gender__c": "Female",
      "Name": "Neomov - Female frame",
      "Price__c": 1450,
    },
    Object {
      "Gender__c": "Female",
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve the products with suspensions 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Neomov - Suspension",
      "Price__c": 1850,
      "Suspension__c": true,
    },
    Object {
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
      "Suspension__c": true,
    },
    Object {
      "Name": "Rolling Mountain - Male",
      "Price__c": 2300,
      "Suspension__c": true,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve the products with the "Rolling Mountain" family Name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "Name": "Rolling Mountain - Female",
      "Price__c": 2300,
      "Product_Family__c": Object {
        "Name": "Rolling Mountain",
      },
    },
  ],
}
`;

exports[`eBikes entity introspection introspect schema 1`] = `
"type Product__c implements Entity {
  Id: ID!
  OwnerId: ID!
  IsDeleted: Boolean!
  Name: String
  CreatedDate: DateTime!
  CreatedById: ID!
  LastModifiedDate: DateTime!
  LastModifiedById: ID!
  SystemModstamp: DateTime!
  LastViewedDate: DateTime
  LastReferencedDate: DateTime
  Autonomy__c: Float
  Description__c: TextArea
  Gender__c: Picklist
  Picture_URL__c: URL!
  Price__c: Currency
  Product_Family__c: Product_Family__c
  Suspension__c: Boolean!
  Order_Items__r(limit: Int!, offset: Int, where: Order_Item__cWhere, order_by: [Order_Item__cOrderBy]): [Order_Item__c]
}

interface Entity {
  Id: ID!
  Name: String
  CreatedDate: DateTime!
  LastModifiedDate: DateTime!
  SystemModstamp: DateTime!
  LastViewedDate: DateTime
}

scalar DateTime

scalar TextArea

scalar Picklist

scalar URL

scalar Currency

input Order_Item__cWhere {
  _and: [Order_Item__cWhere]
  _or: [Order_Item__cWhere]
  Id: IDOperator
  IsDeleted: BooleanOperator
  Name: StringOperator
  CreatedDate: DateTimeOperator
  CreatedById: OrderByEnum
  LastModifiedDate: DateTimeOperator
  LastModifiedById: OrderByEnum
  SystemModstamp: DateTimeOperator
  LastViewedDate: DateTimeOperator
  LastReferencedDate: DateTimeOperator
  Order__c: Order__cWhere
  Product__c: Product__cWhere
}

input IDOperator {
  _eq: ID
  _neq: ID
  _gt: ID
  _lt: ID
  _gte: ID
  _lte: ID
  _in: [ID]
  _nin: [ID]
}

input BooleanOperator {
  _eq: Boolean
  _neq: Boolean
  _gt: Boolean
  _lt: Boolean
  _gte: Boolean
  _lte: Boolean
  _in: [Boolean]
  _nin: [Boolean]
}

input StringOperator {
  _eq: String
  _neq: String
  _gt: String
  _lt: String
  _gte: String
  _lte: String
  _in: [String]
  _nin: [String]
  _like: String
}

input DateTimeOperator {
  _eq: DateTime
  _neq: DateTime
  _gt: DateTime
  _lt: DateTime
  _gte: DateTime
  _lte: DateTime
  _in: [DateTime]
  _nin: [DateTime]
}

enum OrderByEnum {
  ASC
  DESC
  ASC_NULLS_FIRST
  ASC_NULLS_LAST
  DESC_NULLS_FIRST
  DESC_NULLS_LAST
}

input Order__cWhere {
  _and: [Order__cWhere]
  _or: [Order__cWhere]
  Id: IDOperator
  IsDeleted: BooleanOperator
  Name: StringOperator
  CreatedDate: DateTimeOperator
  CreatedById: OrderByEnum
  LastModifiedDate: DateTimeOperator
  LastModifiedById: OrderByEnum
  SystemModstamp: DateTimeOperator
  LastViewedDate: DateTimeOperator
  LastReferencedDate: DateTimeOperator
}

input Product__cWhere {
  _and: [Product__cWhere]
  _or: [Product__cWhere]
  Id: IDOperator
  IsDeleted: BooleanOperator
  Name: StringOperator
  CreatedDate: DateTimeOperator
  CreatedById: OrderByEnum
  LastModifiedDate: DateTimeOperator
  LastModifiedById: OrderByEnum
  SystemModstamp: DateTimeOperator
  LastViewedDate: DateTimeOperator
  LastReferencedDate: DateTimeOperator
  Autonomy__c: FloatOperator
  Gender__c: PicklistOperator
  Picture_URL__c: URLOperator
  Price__c: CurrencyOperator
  Product_Family__c: Product_Family__cWhere
  Suspension__c: BooleanOperator
}

input FloatOperator {
  _eq: Float
  _neq: Float
  _gt: Float
  _lt: Float
  _gte: Float
  _lte: Float
  _in: [Float]
  _nin: [Float]
}

input PicklistOperator {
  _eq: Picklist
  _neq: Picklist
  _gt: Picklist
  _lt: Picklist
  _gte: Picklist
  _lte: Picklist
  _in: [Picklist]
  _nin: [Picklist]
}

input URLOperator {
  _eq: URL
  _neq: URL
  _gt: URL
  _lt: URL
  _gte: URL
  _lte: URL
  _in: [URL]
  _nin: [URL]
}

input CurrencyOperator {
  _eq: Currency
  _neq: Currency
  _gt: Currency
  _lt: Currency
  _gte: Currency
  _lte: Currency
  _in: [Currency]
  _nin: [Currency]
}

input Product_Family__cWhere {
  _and: [Product_Family__cWhere]
  _or: [Product_Family__cWhere]
  Id: IDOperator
  IsDeleted: BooleanOperator
  Name: StringOperator
  CreatedDate: DateTimeOperator
  CreatedById: OrderByEnum
  LastModifiedDate: DateTimeOperator
  LastModifiedById: OrderByEnum
  SystemModstamp: DateTimeOperator
  LastViewedDate: DateTimeOperator
  LastReferencedDate: DateTimeOperator
  Category__c: PicklistOperator
}

input Order_Item__cOrderBy {
  Id: OrderByEnum
  IsDeleted: OrderByEnum
  Name: OrderByEnum
  CreatedDate: OrderByEnum
  CreatedById: OrderByEnum
  LastModifiedDate: OrderByEnum
  LastModifiedById: OrderByEnum
  SystemModstamp: OrderByEnum
  LastViewedDate: OrderByEnum
  LastReferencedDate: OrderByEnum
  Order__c: [Order__cOrderBy]
  Product__c: [Product__cOrderBy]
}

input Order__cOrderBy {
  Id: OrderByEnum
  OwnerId: OrderByEnum
  IsDeleted: OrderByEnum
  Name: OrderByEnum
  CreatedDate: OrderByEnum
  CreatedById: OrderByEnum
  LastModifiedDate: OrderByEnum
  LastModifiedById: OrderByEnum
  SystemModstamp: OrderByEnum
  LastViewedDate: OrderByEnum
  LastReferencedDate: OrderByEnum
}

input Product__cOrderBy {
  Id: OrderByEnum
  OwnerId: OrderByEnum
  IsDeleted: OrderByEnum
  Name: OrderByEnum
  CreatedDate: OrderByEnum
  CreatedById: OrderByEnum
  LastModifiedDate: OrderByEnum
  LastModifiedById: OrderByEnum
  SystemModstamp: OrderByEnum
  LastViewedDate: OrderByEnum
  LastReferencedDate: OrderByEnum
  Autonomy__c: OrderByEnum
  Gender__c: OrderByEnum
  Picture_URL__c: OrderByEnum
  Price__c: OrderByEnum
  Product_Family__c: [Product_Family__cOrderBy]
  Suspension__c: OrderByEnum
}

input Product_Family__cOrderBy {
  Id: OrderByEnum
  OwnerId: OrderByEnum
  IsDeleted: OrderByEnum
  Name: OrderByEnum
  CreatedDate: OrderByEnum
  CreatedById: OrderByEnum
  LastModifiedDate: OrderByEnum
  LastModifiedById: OrderByEnum
  SystemModstamp: OrderByEnum
  LastViewedDate: OrderByEnum
  LastReferencedDate: OrderByEnum
  Category__c: OrderByEnum
}

type Product_Family__c implements Entity {
  Id: ID!
  OwnerId: ID!
  IsDeleted: Boolean!
  Name: String
  CreatedDate: DateTime!
  CreatedById: ID!
  LastModifiedDate: DateTime!
  LastModifiedById: ID!
  SystemModstamp: DateTime!
  LastViewedDate: DateTime
  LastReferencedDate: DateTime
  Category__c: Picklist
  Description__c: TextArea
  Products__r(limit: Int!, offset: Int, where: Product__cWhere, order_by: [Product__cOrderBy]): [Product__c]
}

type Order_Item__c implements Entity {
  Id: ID!
  IsDeleted: Boolean!
  Name: String!
  CreatedDate: DateTime!
  CreatedById: ID!
  LastModifiedDate: DateTime!
  LastModifiedById: ID!
  SystemModstamp: DateTime!
  LastViewedDate: DateTime
  LastReferencedDate: DateTime
  Order__c: Order__c!
  Product__c: Product__c!
}

type Order__c implements Entity {
  Id: ID!
  OwnerId: ID!
  IsDeleted: Boolean!
  Name: String!
  CreatedDate: DateTime!
  CreatedById: ID!
  LastModifiedDate: DateTime!
  LastModifiedById: ID!
  SystemModstamp: DateTime!
  LastViewedDate: DateTime
  LastReferencedDate: DateTime
  Order_Items__r(limit: Int!, offset: Int, where: Order_Item__cWhere, order_by: [Order_Item__cOrderBy]): [Order_Item__c]
}

type Query {
  Product__c(limit: Int!, offset: Int, where: Product__cWhere, order_by: [Product__cOrderBy]): [Product__c]
  Product__c_by_id(id: ID): Product__c
  Product_Family__c(limit: Int!, offset: Int, where: Product_Family__cWhere, order_by: [Product_Family__cOrderBy]): [Product_Family__c]
  Product_Family__c_by_id(id: ID): Product_Family__c
  Order_Item__c(limit: Int!, offset: Int, where: Order_Item__cWhere, order_by: [Order_Item__cOrderBy]): [Order_Item__c]
  Order_Item__c_by_id(id: ID): Order_Item__c
  Order__c(limit: Int!, offset: Int, where: Order__cWhere, order_by: [Order__cOrderBy]): [Order__c]
  Order__c_by_id(id: ID): Order__c
}"
`;
